File: src/main/kotlin/com/huawei/cangjie/psi/CjAnnotationEntry.java
package com.huawei.cangjie.psi;

import com.huawei.cangjie.CjNodeTypes;
import com.huawei.cangjie.name.Name;
import com.huawei.cangjie.psi.stubs.CangJieAnnotationEntryStub;
import com.huawei.cangjie.psi.stubs.elements.CjStubElementTypes;
import com.intellij.lang.ASTNode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class CjAnnotationEntry extends CjElementImplStub<CangJieAnnotationEntryStub> implements CjCallElement  {


    public CjAnnotationEntry(@NotNull ASTNode node) {
        super(node);
    }

    public CjAnnotationEntry(@NotNull CangJieAnnotationEntryStub stub) {
        super(stub, CjStubElementTypes.ANNOTATION_ENTRY);
    }
    @Override
    public CjConstructorCalleeExpression getCalleeExpression() {
        return getStubOrPsiChild(CjStubElementTypes.CONSTRUCTOR_CALLEE);
    }
    @Override
    public CjValueArgumentList getValueArgumentList() {
        CangJieAnnotationEntryStub stub = getStub();
        if (stub == null && getGreenStub() != null) {
            return (CjValueArgumentList) findChildByType(CjNodeTypes.VALUE_ARGUMENT_LIST);
        }

        return getStubOrPsiChild(CjStubElementTypes.VALUE_ARGUMENT_LIST);
    }
    @Nullable @IfNotParsed
    public CjTypeReference getTypeReference() {
        CjConstructorCalleeExpression calleeExpression = getCalleeExpression();
        if (calleeExpression == null) {
            return null;
        }
        return calleeExpression.getTypeReference();
    }
    @Nullable
    public Name getShortName() {
      CangJieAnnotationEntryStub stub = getStub();
        if (stub != null) {
            String shortName = stub.getShortName();
            if (shortName != null) {
                return Name.identifier(shortName);
            }
            return null;
        }

        CjTypeReference typeReference = getTypeReference();
        assert typeReference != null : "Annotation entry hasn't typeReference " + getText();
       CjTypeElement typeElement = typeReference.getTypeElement();
        if (typeElement instanceof CjUserType userType) {
            String shortName = userType.getReferencedName();
            if (shortName != null) {
                return Name.identifier(shortName);
            }
        }
        return null;
    }


}

File: src/main/kotlin/com/huawei/cangjie/psi/CjAnnotationEntry.kt
package com.huawei.cangjie.psi

import com.huawei.cangjie.CjNodeTypes
import com.huawei.cangjie.name.Name
import com.huawei.cangjie.name.Name.Companion.identifier
import com.huawei.cangjie.psi.stubs.CangJieAnnotationEntryStub
import com.huawei.cangjie.psi.stubs.elements.CjStubElementTypes
import com.intellij.lang.ASTNode

class CjAnnotationEntry : CjElementImplStub<CangJieAnnotationEntryStub >, CjCallElement {
    constructor(node: ASTNode) : super(node)

    constructor(stub: CangJieAnnotationEntryStub) : super(stub, CjStubElementTypes.ANNOTATION_ENTRY)

    override val calleeExpression : CjConstructorCalleeExpression? get(){
        return getStubOrPsiChild(CjStubElementTypes.CONSTRUCTOR_CALLEE)
    }

    override val lambdaArguments: List<CjLambdaArgument> get(){
        return emptyList()
    }

    override val typeArguments : List<CjTypeProjection> get(){
        val typeArgumentList = typeArgumentList ?: return emptyList()
        return typeArgumentList.arguments
    }

    override val typeArgumentList : CjTypeArgumentList? = null

    override val valueArgumentList : CjValueArgumentList? get(){
        val stub = stub
        if (stub == null && greenStub != null) {
            return findChildByType(CjNodeTypes.VALUE_ARGUMENT_LIST)
        }

        return getStubOrPsiChild(CjStubElementTypes.VALUE_ARGUMENT_LIST)
    }

    override val valueArguments : List<ValueArgument > get(){
        val stub = stub
        if (stub != null && !stub.hasValueArguments()) {
            return emptyList<CjValueArgument>()
        }

        val list = valueArgumentList
        return list?.arguments ?: emptyList<CjValueArgument>()
    }

    @get:IfNotParsed
    val typeReference: CjTypeReference?
        get() {
            val calleeExpression = calleeExpression ?: return null
            return calleeExpression.typeReference
        }
    val shortName: Name?
        get() {
            val stub = stub
            if (stub != null) {
                val shortName = stub.getShortName()
                if (shortName != null) {
                    return identifier(shortName)
                }
                return null
            }

            val typeReference =
                checkNotNull(typeReference) { "Annotation entry hasn't typeReference $text" }
            val typeElement = typeReference.typeElement
            if (typeElement is CjUserType) {
                val shortName = typeElement.referencedName
                if (shortName != null) {
                    return identifier(shortName)
                }
            }
            return null
        }

}
